### PR TITLE
Singleton design pattern Code

### PR DESCRIPTION
**Pull request by JotsnaW**
Created simple Singleton design pattern Code

### CODE REVIEW

File: SingletonDesignPattern/Properties/AssemblyInfo.cs
Comment (MohdSarvej) [inline SingletonDesignPattern/Properties/AssemblyInfo.cs:1]: File not needed.

Comment (JotsnaW) [reply SingletonDesignPattern/Properties/AssemblyInfo.cs:1]: deleted the file

File: SingletonDesignPattern/Program.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+
+namespace SingletonDesignPattern
+{
+    internal class Program
+    {
+        static void Main(string[] args)
+        {
+            // call to get the instance of Singleton class
+            Singleton SingletonInstance1 = Singleton.GetInstance();
```
Comment (asadhasnain) [inline SingletonDesignPattern/Program.cs]: Better naming is the key for clean code. Please review the rest of the code too

File: MultiLoggerSingleton2/Logger.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        public static Logger GetInstance(string fileName)
+        {
+            lock (lockObject)
+            {
+                if (!m_object.ContainsKey(fileName))
+                {
+                    m_object.Add(fileName, new Logger());
+                }
+            }
+            m_object.TryGetValue(fileName, out Logger instance);
```
Comment (asadhasnain) [inline MultiLoggerSingleton2/Logger.cs]: If you do this check first then if not exist then create the logger and Add it.

reason: Think of logger, the log file instance is created once and reused again and again.

File: MultiLoggerSingleton2/Program.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        /// <summary>
+        /// This Program is an example of Thread Safe Singleton - MultiLogger with different files.
+        /// </summary>
+        static void Main()
+        {
+            string fileName1 = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\log1.text";
+            string fileName2 = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\log2.text";
+
+
+            Logger log1 = Logger.GetInstance(fileName1);
```
Comment (asadhasnain) [inline MultiLoggerSingleton2/Program.cs]: naming is the key in any solution.

Logs name change be: NetworkTraffic.log
Database.log
Error.log
UserAccess.log

File: MultiLoggerSingleton2/Program.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+namespace MultiLoggerSingleton2
+{
+    internal class Program
+    {
+        /// <summary>
+        /// This Program is an example of Thread Safe Singleton - MultiLogger with different files.
+        /// </summary>
+        static void Main()
+        {
+            string fileName1 = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\log1.text";
```
Comment (asadhasnain) [inline MultiLoggerSingleton2/Program.cs]: text is never any extension.

Either it should be txt or log

File: MultiLoggerSingleton/Logger.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.IO;
+
+namespace MultiLoggerSingleton
+{
+    public class Logger: ILogger
+    {
+        public static int m_count = 0;
```
Comment (asadhasnain) [inline MultiLoggerSingleton/Logger.cs]: What benefit this m_count will do?

File: MultiLoggerSingleton/Logger.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        { 
+            m_count++;
+            Console.WriteLine($"Constructor is called for {m_count} time.");
+            m_logFile = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\logFile.txt";
+        }
+
+        // static method to get instance
+        public static Logger GetInstance()
+        {
+            if(instance == null)
```
Comment (asadhasnain) [inline MultiLoggerSingleton/Logger.cs]: do we really need this?

File: MultiLoggerSingleton/Logger.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        private readonly string m_logFile;
+        private static Logger instance = null;
+        private static readonly object lockobject  = new object();
+
+        // private constructor
+        private Logger() 
+        { 
+            m_count++;
+            Console.WriteLine($"Constructor is called for {m_count} time.");
+            m_logFile = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\logFile.txt";
```
Comment (asadhasnain) [inline MultiLoggerSingleton/Logger.cs]: Hard coded path

File: MultiLoggerSingleton/Program.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+namespace MultiLoggerSingleton
+{
+    internal class Program
+    {
+        /// <summary>
+        /// This Program is an example of Thread Safe Singleton - MultiLogger with single files.
+        /// </summary>
+        static void Main()
+        {
+            Logger logger1 = Logger.GetInstance();
```
Comment (asadhasnain) [inline MultiLoggerSingleton/Program.cs]: Better language

File: MultiLoggerSingleton2/Logger.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    public class Logger: ILogger
+    {
+        public static int m_count = 0;
+        private static readonly object lockObject = new object();
+        private static Dictionary<string, Logger> m_object = new Dictionary<string, Logger>();
+
+        // private constructor
+        private Logger()
+        {
+            m_count++;
```
Comment (asadhasnain) [inline MultiLoggerSingleton2/Logger.cs]: Why we need this m_count

File: SingletonDesignPattern/Singleton.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            return instance;
+        }
+
+        public string getValue()
+        {
+            return name;
+        }
+
+        public void setValue(string text)
```
Comment (asadhasnain) [inline SingletonDesignPattern/Singleton.cs]: should be Pascal right?

**Review by asadhasnain — COMMENTED** [review]
Added comments for your consideration

File: SingletonDesignPattern/Singleton.cs (hunk 10 above)
Comment (JotsnaW) [reply SingletonDesignPattern/Singleton.cs]: Yes, method names should be in PascalCase. Updated method names GetValue( ) and SetValue( ).

File: SingletonDesignPattern/Program.cs (hunk 1 above)
Comment (JotsnaW) [reply SingletonDesignPattern/Program.cs]: Updated instance names and reviewed other code for the same.

File: MultiLoggerSingleton2/Program.cs (hunk 4 above)
Comment (JotsnaW) [reply MultiLoggerSingleton2/Program.cs]: Updated

File: MultiLoggerSingleton2/Logger.cs (hunk 9 above)
Comment (JotsnaW) [reply MultiLoggerSingleton2/Logger.cs]: just an extra check for Singleton. 
I used this variable to check how many times our constructor is getting called.

File: MultiLoggerSingleton/Logger.cs (hunk 6 above)
Comment (JotsnaW) [reply MultiLoggerSingleton/Logger.cs]: removed extra null check

File: MultiLoggerSingleton/Logger.cs (hunk 5 above)
Comment (JotsnaW) [reply MultiLoggerSingleton/Logger.cs]: Added while writing code to check that constructor is getting called only once for Singleton.
I will remove it.

File: MultiLoggerSingleton/Logger.cs (hunk 7 above)
Comment (JotsnaW) [reply MultiLoggerSingleton/Logger.cs]: removed hard coded path from Logger class and added it in Main( ) and passing as argument to LogToFile ( ).

File: MultiLoggerSingleton/Program.cs (hunk 8 above)
Comment (JotsnaW) [reply MultiLoggerSingleton/Program.cs]: Updated.

File: MultiLoggerSingleton2/Program.cs (hunk 3 above)
Comment (JotsnaW) [reply MultiLoggerSingleton2/Program.cs]: Updated

File: MultiLoggerSingleton2/Logger.cs (hunk 2 above)
Comment (JotsnaW) [reply MultiLoggerSingleton2/Logger.cs]: yes, right. 
Updated if statement check.

File: MultiLoggerSingleton2/Logger.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+                if ( !m_objects.TryGetValue(LogFileName, out Logger instance) )
+                    {
+                    m_objects.Add(LogFileName, instance = new Logger());
+                    }
+                return instance;
+                }
+            }
+
+        public void LogToFile(string LogFileName, string logMessage)
+        {
```
Comment (asadhasnain) [inline MultiLoggerSingleton2/Logger.cs]: extra level of complexity for caller

File: MultiLoggerSingleton2/Program.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            string DatabaseLogFile = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\Database.log";
+            string ErrorLogFile = "D:\\GitHub\\DesignPatterns\\MultiLoggerSingleton\\Error.log";
+
+
+            Logger firstLoggerObject = Logger.GetLoggerInstance(DatabaseLogFile);
+            firstLoggerObject.LogToFile(DatabaseLogFile, "Database Log entry");
+
+            Logger secondLoggerObject = Logger.GetLoggerInstance(ErrorLogFile);
+            secondLoggerObject.LogToFile(ErrorLogFile, "Error Log entry");
```
Comment (asadhasnain) [inline MultiLoggerSingleton2/Program.cs]: there shouldn't be any need to pass the log file path with each log request

**Review by asadhasnain — COMMENTED** [review]
Added comments please review

**Review by asadhasnain — APPROVED** [review]
Approved.

Just fix the names

**Review by MohdSarvej — APPROVED** [review]

**Pull request merged**